### PR TITLE
[LEVWEB-1393] redact death data

### PR DESCRIPTION
**Pull request by aarongill1**

### CODE REVIEW

File: test/unit/lib/redaction.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+const roleWithoutFullAccess = ['births'];
+const deathData = [{
+  date: 'date',
+  dataToBeRedacted1: 'extraData',
+  deceased: {
+    forenames: 'forenames',
+    surname: 'surname',
+    dataToBeRedacted2: 'moreData'
+  }
+}]
```
Comment (daniel-ac-martin) [inline test/unit/lib/redaction.js:14]: This should be an object not an array with one element.

File: src/middleware/v1/registration/death.js [hunk 2]
```diff
@@ -66,7 +67,7 @@ module.exports = {
         audit.create(ri.username, ri.client, req.url, ri.groups, 'search', 'death')
           .then(() => model.search(query))
           .then(r => {
-            res.send(r);
+            res.send(redactDeath(r, ri.roles));
```
Comment (daniel-ac-martin) [inline src/middleware/v1/registration/death.js:70]: We should map `redactDeath` over `r`.

```
res.send(r.map(e => redactDeath(e, ri.roles)));
```

Comment (daniel-ac-martin) [reply src/middleware/v1/registration/death.js:70]: Or something like that.

Comment (daniel-ac-martin) [reply src/middleware/v1/registration/death.js:70]: We should also put the existing version in to the `read` function earlier in the file. I forgot to mention that.

File: src/lib/redaction.js [hunk 3]
```diff
@@ -0,0 +1,12 @@
+const redactDeath = (death, roles) => {
+  return roles.includes('full-details') ? death :
+    [{
+      date: death[0].date,
+      deceased: {
+        forenames: death[0].deceased.forenames,
+        surname: death[0].deceased.surname
+      }
+    }]
```
Comment (daniel-ac-martin) [inline src/lib/redaction.js:9]: As discussed in the test file. This function should take and return an object not and array.

To run it on an array we can just map it over the array.

File: src/lib/redaction.js [hunk 4]
```diff
@@ -1,10 +1,68 @@
...
+        retired: death.deceased.retired,
+        maidenSurname: death.deceased.maidenSurname,
+        aliases: death.deceased.aliases.map((e) => {
+          return {
+            type: e.type,
+            prefix: e.prefix,
+            forenames: e.forenames,
+            surname: e.surname,
+            suffix: e.suffix
+          }
```
Comment (daniel-ac-martin) [inline src/lib/redaction.js:32]: Is this done simply to clone the object?

If so, I don't think you really need to clone. - I'd probably just assign the existing object like everything else. (@RobinKnipe: Any thoughts on this?)

If you did still to clone there are better ways, either with `Object.assign` or spread syntax, i.e:
```js
aliases: { ...death.deceased.aliases }
```

File: src/lib/redaction.js [hunk 5]
```diff
@@ -1,10 +1,68 @@
...
+        signature: death.registrar.signature,
+        designation: death.registrar.designation,
+        subdistrict: death.registrar.subdistrict,
+        district: death.registrar.district,
+        administrativeArea: death.registrar.administrativeArea
+      },
+      coroner: {
+        name: death.coroner.name,
+        designation: death.coroner.designation,
+        area: death.coroner.area
```
Comment (daniel-ac-martin) [inline src/lib/redaction.js:65]: This looks like too many fields. Are you intended to reduce this to only the mandatory ones in the next commit? (I'd keep this in the commit history as it's a useful reference.)

File: src/lib/redaction.js (hunk 4 above)
Comment (daniel-ac-martin) [reply src/lib/redaction.js:32]: Sorry just realised you've got an array so you would use:
```js
aliases: [ ...death.deceased.aliases ],
```
If you wanted to clone the array.

Comment (aarongill1) [reply src/lib/redaction.js:32]: Yeah, not really sure what I was thinking here, changed to:

aliases: death.deceased.aliases

**Review by RobinKnipe — CHANGES_REQUESTED** [review]
A couple of things to look at... but it's basically there

File: test/unit/lib/redaction.js [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+  }
+}
+
+describe('lib/redaction.js', () => {
+  describe('redactDeath function', () => {
+    it('should be a function', () => {
+      expect(redactDeath).to.be.a('function')
+    })
+    it('should return the death data if role has full access', () => {
+      expect(redactDeath(deathData, roleWithFullAccess)).to.deep.equal(deathData);
```
[the remark this answers is not in the document]
Comment (aarongill1) [reply test/unit/lib/redaction.js:171]: Done

File: test/unit/lib/redaction.js [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+    })
+    it('should return redacted death data if role has not got full access', () => {
+      expect(redactDeath(deathData, roleWithoutFullAccess)).to.deep.equal(redactedData);
```
[the remark this answers is not in the document]
Comment (aarongill1) [reply test/unit/lib/redaction.js:174]: Added a test for this on a random property. Also added a test for an exact key match - if there is anything else, the test fails.

File: src/lib/redaction.js (hunk 5 above)
Comment (aarongill1) [reply src/lib/redaction.js:65]: Sorted

File: src/lib/redaction.js [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+        causeOfDeath: death.deceased.causeOfDeath,
+        certifiedBy: death.deceased.certifiedBy,
+        sex: death.deceased.sex,
+        address: death.deceased.address,
+        occupation: death.deceased.occupation,
+        relationshipToPartner: death.deceased.relationshipToPartner,
+        retired: death.deceased.retired,
+        maidenSurname: death.deceased.maidenSurname,
+        aliases: death.deceased.aliases.map((e) => {
+          return {
```
[the remark this answers is not in the document]
Comment (aarongill1) [reply src/lib/redaction.js:26]: Improved as per previous suggested but good tip!

**Comment by daniel-ac-martin** [comment]
This is looking pretty good to me.

We should bump the version somehow. Probably the minor version.

We should also have a separate PR on the lev-api-spec repo to point out to users that they will only receive fields that they have permission to see regardless of whether the field is marked as required or not.

**Review by RobinKnipe — DISMISSED** [review]
What Daniel said ^. LGTM

**Review by daniel-ac-martin — CHANGES_REQUESTED** [review]
Just bump the minor version.

**Comment by daniel-ac-martin** [comment]
Also we should get the lev-api-spec change in place first. (Along with its own version bump.)

**Review by daniel-ac-martin — APPROVED** [review]

**Pull request merged**